src/fastertransformer/devices/rocm_impl/ROCmAllocator.h

```diff
@@ -1,32 +1,51 @@
 #pragma once
+
 #include "src/fastertransformer/core/allocator.h"
+#include "src/fastertransformer/devices/rocm_impl/ROCmUtils.h"
+#include <hip/hip_runtime.h>
 
 namespace fastertransformer {
 
-template<>
-class Allocator<AllocatorType::ROCM> : public TypedAllocator<AllocatorType::ROCM> {
+template<AllocatorType AType, MemoryType MType, hipError_t (*Alloc)(void**, size_t), hipError_t (*Free)(void*)>
```

Comment:
why would you add function signatures in template arguments?

Reply:
I think that is the only way for allowing functions as template arguments prior to c++17?

Reply:
But why do you need to add functions as template arg?

Reply:
and, I think we are using c++17.

Reply:
Just I could easily specialize device and host allocators. Since they are the same just using different HIP API functions.